Add a FIPS provider and implement SHA256 in it

mattcaswell opened the pull request:
This PR is built on top of #8513. It creates the FIPS provider and implements SHA256 within it.

So far this only includes the C implementation of SHA256. Adding the assembler modules is a bit trickier so that is deferred for a later time. A consequence of that is that (AFAICT) OPENSSL_cleanse is only implemented in assembler so (for now) I have implemented a very simple C version of OPENSSL_cleanse. It's totally unsafe to use in production so will need to be removed before too long, but it gets things working for now.

I encountered some unexpected behaviour during the implementation of this. At one point I had created the FIPS provider but not yet added all the files I needed for SHA256. Therefore I was expecting the test I had written to use the new provider to fail due to some symbols not being present. To my surprise the test passed - the module loaded and performed SHA256 sucessfully. It seems that, when building the FIPS provider, if there are symbols from libcrypto that it needs but that aren't present, those symbols are resolved automatically if the provider is loaded by an application linked against libcrypto. This is not desirable behaviour since the FIPS module must be entirely self-contained. To detect this problem I created a new test in shlibloadtest that simply loads the FIPS module from an application that isn't linked with libcrypto. If there are no missing symbols then this should work and will fail otherwise.

Finally, this PR also implements a "no-fips" Configure option.

levitte commented:
> A consequence of that is that (AFAICT) OPENSSL_cleanse is only implemented in assembler so (for now) I have implemented a very simple C version of OPENSSL_cleanse.

[ahem] `crypto/mem_clr.c`

levitte commented:
> It seems that, when building the FIPS provider, if there are symbols from libcrypto that it needs but that aren't present, those symbols are resolved automatically if the provider is loaded by an application linked against libcrypto.

I wonder if this is true on all platforms...  I *think* that Windows demands that all symbols be resolved when linking the DLL, and I know for a fact that this is the case on VMS.

mattcaswell commented:
Darn. I looked everywhere for that and couldn't find it!! Thanks for the pointer.

levitte commented:
`git grep` is your friend :wink:

mattcaswell commented:
> I wonder if this is true on all platforms... I think that Windows demands that all symbols be resolved when linking the DLL, and I know for a fact that this is the case on VMS.

I could imagine that being highly platform specific

mattcaswell commented:
> git grep is your friend wink

Too many hits to find the right one

levitte commented:
> > I wonder if this is true on all platforms... I think that Windows demands that all symbols be resolved when linking the DLL, and I know for a fact that this is the case on VMS.
> 
> I could imagine that being highly platform specific

I believe that `-z defs` may be the answer...  I wonder if we should have something like that for all our modules, engines and providers alike...

levitte commented on crypto/build.info:
```diff
@@ -20,6 +20,9 @@ SOURCE[../libcrypto]=\
 
+SOURCE[../providers/fips]=\
+        cryptlib.c mem.c
+
```
I would much rather you place these in `providers/build.info`, like this:

```
SOURCE[fips]=../crypto/cryptlib.c ../crypto/mem.c
```

levitte commented on crypto/sha/build.info:
```diff
@@ -3,6 +3,9 @@ SOURCE[../../libcrypto]=\
         {- $target{keccak1600_asm_src} -}
 
+SOURCE[../../providers/fips]=\
+        sha256.c
```
Likewise, in `providers/build.info`:

```
SOURCE[fips]=../crypto/sha/sha256.c
```

mattcaswell commented:
I updated this to address comments so far and rebased to pull in the commits from #8513.

I dropped the changes to shlibloadtest because I found it to be unreliable. Instead I've amended the symbol_presence test to check for libcrypto symbols in fips.so that shouldn't be there.

I've also taken this out of WIP.

paulidale commented on providers/fips/fipsprov.c:
```diff
@@ -0,0 +1,99 @@
+    const OSSL_PARAM *p;
+
+    p = OSSL_PARAM_locate(params, OSSL_PROV_PARAM_NAME);
+    if (p != NULL && !OSSL_PARAM_set_utf8_ptr(p, "OpenSSL Default Provider"))
```
FIPS

paulidale commented on test/evp_extra_test.c:
```diff
@@ -1086,26 +1093,61 @@ static int test_EVP_MD_fetch(int tst)
-            || !TEST_true(calculate_digest(md, testmsg, sizeof(testmsg), exptd)))
-        goto err;
+    /*
+     * Explicitly asking for the default implementation should succeeed except
```
succeed
(which I missed on the earlier PR)

levitte commented on test/recipes/01-test_symbol_presence.t:
```diff
@@ -29,10 +36,21 @@ note
+            $libname = "crypto";
+        } else {
+            $libname = $thislib;
+            $shlibpath = bldtop_file("lib" . $libname . ".so");
```
This could use `platform->shlib`...

levitte commented:
We could really make sure that all modules are linked with all symbols resolved. This requires adding this appropriately in `Configurations/shared_info.pl`:

```
        dso_ldflags => '-z defs', 
```

... or similar depending on the platform 

It does make sense to do this for engines as well, really, and they should get an additional `-lcrypto` if they don't already have that

mattcaswell commented:
Fixups pushed addressing the feedback so far. The "-z defs" change kind of makes the symbol_presence checks redundant...but that only impacts linux so maybe its useful still on some platforms? I've left it in for now.

levitte commented on Configurations/shared-info.pl:
```diff
@@ -32,6 +32,7 @@ sub detect_gnu_cc {
         return {
             %{$shared_info{'gnu-shared'}},
             shared_defflag    => '-Wl,--version-script=',
+            dso_ldflags => '-z defs',
```
Nit: It would be nice if `=>` was aligned with the previous line...

mattcaswell replied:
Done

levitte commented:
> Fixups pushed addressing the feedback so far. The "-z defs" change kind of makes the symbol_presence checks redundant...but that only impacts linux so maybe its useful still on some platforms? I've left it in for now.

Good idea.  That needs a more general fixup for other Unixen, and possibly for Windows

mattcaswell commented:
Hmmm....I have no idea what to do about this travis error:
````
Unknown operating system family darwin

make[1]: *** [providers/fips.ld] Error 255

make: *** [all] Error 2

+//// MAKE FAILED
````

mattcaswell commented:
Ah! This error comes from mkdef.pl:

https://github.com/openssl/openssl/blob/b3d113ed2993801ee643126118ccf6592ad18ef7/util/mkdef.pl#L109-L143

And appears to come ultimately from the "shared_target" value in the "Configurations" files. If that's the case it looks like there are quite a number missing from the list in mkdef.pl. Unfortunately I'm not sure what they should be set to for the missing platforms....@levitte?

levitte commented:
Hmmm... mkdef.pl should check if `$target{shared_defflag}` is defined, and simply do nothing if it isn't (except exit 0).  I think...
Remind me on Tuesday if you don't figure it out on your own.

paulidale commented:
@levitte late reminder :)

levitte commented on providers/build.info:
```diff
@@ -1 +1,13 @@
+  SUBDIRS=fips
+  MODULES=fips
+  SHARED_SOURCE[fips]=fips.ld
+  GENERATE[fips.ld]=provider.num
```
These kinds of lines is usually wrapped like this:

```
  IF[{- defined $target{shared_defflag} -}]
    SHARED_SOURCE[fips]=fips.ld
    GENERATE[fips.ld]=provider.num
  ENDIF
```

That's easily forgotten, though, so maybe Configure should be adjusted to ignore any `.ld` file spec if `$target{shared_defflag}` isn't defined...  it seems a bit "magic" for my taste, but...

levitte replied:
The lack of this guard is what causes the `Unknown operating system family darwin` issue.

mattcaswell replied:
Ah, thanks! Will fix. That was non-obvious!

mattcaswell replied:
Fixup pushed. Lets see if travis is happier.

mattcaswell commented:
Travis still failed, although this time only in one build. In that build the symbol_presence test failed - complaining that *all* libcrypto symbols were absent. I don't see how this PR could have caused that (especially since all the other tests passed successfully) - so I suspect an anomaly. The appveyor failure *was* an issue and I have pushed a fixup for that. Lets see what happens this time around.

mattcaswell commented:
Appveyor passed this time but Travis still failed :-(

I really have no clue as to why the symbol_presence test is failing in this one build.

@levitte - any ideas?

levitte commented on test/recipes/01-test_symbol_presence.t:
```diff
@@ -29,10 +36,21 @@ note
+            skip "FIPS is disabled in this build", 2 if disabled("fips");
+            $shlibpath = bldtop_file("providers", platform->dso('fips'));
+            #Actually we're going to compare it against libcrypto
+            $libname = "crypto";
```
Uhmmm, no we are not.  We should compare against `util/providers.num`  Please remember that we're checking *exported* symbols.

levitte commented on test/recipes/01-test_symbol_presence.t:
```diff
@@ -9,15 +9,22 @@
     if disabled("shared");
 
-my @libnames = ("crypto", "ssl");
+my @libnames = ("crypto", "ssl", "fips");
```
This list could really be refactored to:

``` perl
my @libnames = ("libcrypto", "libssl", "fips");
```

That way, you don't have to add "lib" all over the place.

Also, it might be prudent to make this an associative list, to map libraries/modules against their corresponding ordinals file, so maybe something like this?

``` C
my %libnames = (
    libcrypto => 'libcrypto',
    libssl    => 'libssl',
    fips      => 'provider'
);
```

And then rework the code to use that table to figure out the correct ordinals file to compare with.

levitte commented:
On second thought, the `test/recipes/01-test_symbol_presence.t ` changes aren't terribly useful.  After all, the only thing you get to do with that is to check that the fips module exports `OSSL_provider_init`, and that's terribly exciting...  Woohoo.  I suspect that *other* stuff will break if that single symbol isn't exported.

mattcaswell commented:
Thanks @levitte. I have removed the offending commit and rebased this. Lets see what the CIs think now.

levitte commented on providers/provider.num:
```diff
@@ -0,0 +1 @@
+OSSL_provider_init                     1	*	EXIST::FUNCTION:
```
Wait... Isn't there one in util already?

paulidale replied:
It looks like it, in `util/providers.num`

levitte commented on util/shlib_wrap.sh.in:
```diff
@@ -117,6 +117,9 @@ if [ -f "$LIBCRYPTOSO" -a -z "$preload_var" ]; then
 
+OPENSSL_MODULES="${THERE}/providers"
+export OPENSSL_MODULES
+
```
The trouble with this is that we may fool ourselves into relying on this, which is harmful on non-Unix platforms.

mattcaswell replied:
I don't know about you, but I use this script myself all the time. I think it should just do "the right thing" without having to setup other env variables manually.

levitte replied:
I usually do things like `OPENSSL_MODULES=provider. /util/shlib_wrap.sh apps/openssl ...`

And look, the thing that made med think of this is the assignment of `OPENSSL_MODULES` in a test recipe, which is completely unnecessary in a Unix environment with the change you're making here. Come the day when we've become comfortable with this, and get annoyingly reminded by appveyor that oops, forgot.

Furthermore, there will be times when we may want to use this *without* getting `OPENSSL_MODULES` defined, to check some behavior bug for that scenario.

mattcaswell replied:
> I usually do things like OPENSSL_MODULES=provider ./util/shlib_wrap.sh apps/openssl ...

That's going to become a big pain very quickly. Unlike with engines which we use relatively infrequently I think we're going to be doing this *all* the time.

> And look, the thing that made med think of this is the assignment of OPENSSL_MODULES in a test recipe, which is completely unnecessary in a Unix environment with the change you're making here. Come the day when we've become comfortable with this, and get annoyingly reminded by appveyor that oops, forgot.

Perhaps the issue here is that the script is being used for 2 different things. One as a wrapper for the tests, and one for developers from the command line. Perhaps we need a new script.

> Furthermore, there will be times when we may want to use this without getting OPENSSL_MODULES defined, to check some behavior bug for that scenario.

Well that one is easy. Given that's going to be very infrequent compared to the default case where we do want OPENSSL_MODULES defined - just don't use the script in those rare cases!

Anyway this whole issue is somewhat tangential to this PR. I think we need this for command line usage - but it isn't necessary for this particular PR. Since it is controversial I have removed it for now (I might bring it back in some future PR).

levitte replied:
> Well that one is easy. Given that's going to be very infrequent compared to the default case where we do want OPENSSL_MODULES defined - just don't use the script in those rare cases!

... 'cause any of the newly built programs are doing so well without the newly built libraries. 

> Perhaps we need a new script.

Yes, that's the better idea! 

> Since it is controversial I have removed it for now

Okie, cool

mattcaswell commented:
Due to conflicts with master I rebased this as well as fixing the shlib_wrap issue above.

paulidale commented on providers/fips/fipsprov.c:
```diff
@@ -0,0 +1,99 @@
+    return 1;
+}
+
+extern const OSSL_DISPATCH sha256_functions[];
```
This isn't going to scale but is fine for now.

mattcaswell replied:
You mean "scale" in terms of the number of different algorithms we support? Yeah - we'll probably have a header file I guess for these.

paulidale commented on providers/fips/fipsprov.c:
```diff
@@ -0,0 +1,99 @@
+extern const OSSL_DISPATCH sha256_functions[];
+
+static const OSSL_ALGORITHM fips_digests[] = {
+    { "SHA256", "fips=yes", sha256_functions },
```
"fips=yes" or just "fips" ?

A decision either way will be required at some point.

mattcaswell replied:
Personally I prefer "fips=yes"

paulidale commented on providers/fips/fipsprov.c:
```diff
@@ -0,0 +1,99 @@
+            c_get_params = OSSL_get_core_get_params(in);
+            break;
+        default:
+            /* Just ignore anything we don't understand */
```
On the previous line?

mattcaswell replied:
Done

levitte commented on providers/build.info:
```diff
@@ -1 +1,15 @@
+  MODULES=fips
+  IF[{- defined $target{shared_defflag} -}]
+    SHARED_SOURCE[fips]=fips.ld
+    GENERATE[fips.ld]=provider.num
```
That should be

```
    GENERATE[fips.ld]=../util/providers.num
```

There's no reason to make multiple copies of it...

levitte replied:
("it" being that ordinals file)

mattcaswell replied:
Fixed

mattcaswell commented:
All comments responded to and fixups pushed.

levitte commented:
Travis is acting up.

levitte commented:
In the Appveyor log for x64, I find this:

```
[00:06:05]         # ERROR: (ptr) 'fipsprov != NULL' failed @ ..\test\evp_extra_test.c:1128
[00:06:05]         # 0x0
[00:06:05]         # 2480:error:2506506A:DSO support routines:win32_bind_func:could not bind to the requested symbol name:..\crypto\dso\dso_win32.c:184:symname(OSSL_provider_init)
[00:06:05]         # 2480:error:2506C06A:DSO support routines:DSO_bind_func:could not bind to the requested symbol name:..\crypto\dso\dso_lib.c:186:
[00:06:05]         # 2480:error:0F086046:common libcrypto routines:provider_activate:init fail:..\crypto\provider_core.c:405:name=fips
[00:06:05]         not ok 4 - iteration 4
[00:06:05]         # ERROR: (ptr) 'fipsprov != NULL' failed @ ..\test\evp_extra_test.c:1128
[00:06:05]         # 0x0
[00:06:05]         # 2480:error:2506506A:DSO support routines:win32_bind_func:could not bind to the requested symbol name:..\crypto\dso\dso_win32.c:184:symname(OSSL_provider_init)
[00:06:05]         # 2480:error:2506C06A:DSO support routines:DSO_bind_func:could not bind to the requested symbol name:..\crypto\dso\dso_lib.c:186:
[00:06:05]         # 2480:error:0F086046:common libcrypto routines:provider_activate:init fail:..\crypto\provider_core.c:405:name=fips
[00:06:05]         not ok 5 - iteration 5
[00:06:05]     not ok 13 - test_EVP_MD_fetch
[00:06:05] ..\evp_extra_test.exe => 1
[00:06:05] not ok 1 - running evp_extra_test
```

I tried doing a similar build on my home Windows system (running Windows 10), but couldn't reproduce, the test runs cleanly...

mattcaswell commented:
I couldn't initially reproduce this issue. However, I then tried rebasing my branch on top of a more recent master (which is what AppVeyor does before running tests) and this time I was able to reproduce it. So I suspect that some recent commit to master is the culprit (or at least is related).

levitte commented on providers/build.info:
```diff
@@ -1 +1,15 @@
+  SUBDIRS=fips
+  MODULES=fips
+  IF[{- defined $target{shared_defflag} -}]
+    SHARED_SOURCE[fips]=fips.ld
```
Change `SHARED_SOURCE` to `SOURCE`.  `SHARED_SOURCE` is only meant for libraries, where some sources are only used for the shared variant.

mattcaswell replied:
Fixed

levitte commented:
Oh, and please please rebase

mattcaswell commented:
Fixed the issue above and rebased

levitte commented:
So now that we tamed appveyor, Travis takes its turn...

levitte commented:
Ah, found why:

```
EVP function code 229 collision at EVP_F_EVP_MD_BLOCK_SIZE
```

slontis commented on test/evp_extra_test.c:
```diff
@@ -1207,6 +1250,10 @@ int setup_tests(void)
     ADD_ALL_TESTS(test_invalide_ec_char2_pub_range_decode,
                   OSSL_NELEM(ec_der_pub_keys));
 #endif
+#ifdef OPENSSL_NO_FIPS
```
There is a FIPS_MODE define...

mattcaswell replied:
> There is a FIPS_MODE define...

Yes, I know. It's this PR that ensures it gets defined! (See providers/build.info)

However that define is intended for files that get compiled twice: once for the default provider, and once for the FIPS provider. Whether it is defined or not doesn't tell you anything about whether FIPS has been disabled for this build via "no-fips". For that we need to use OPENSSL_NO_FIPS - as is done here.

paulidale replied:
What do we do about the unit tests?

I've tried to write them so that the mode doesn't matter when compiled but that's not the same as running them twice, once with and once without, some FIPS mode flag set.  I'd assumed we compile the tests twice.

It would be easy enough to switch over to an environment variable or some other conditional.

levitte replied:
For unit tests of FIPS code specifically, it would probably be best to have separate internals test programs to run those. They could be built with `FIPS_MODE` always defined and directly using the source they're supposed to test instead of getting it from libcrypto.

As environment variables go, #8663 might provide an answer.

mattcaswell commented:
Close/reopen to kick CIs

levitte commented on Configure:
```diff
@@ -512,6 +513,8 @@ my @disable_cascades = (
     # or modules.
     "pic"               => [ "shared", "module" ],
 
+    "dynamic-engine"    => [ "fips" ],
```
No.  If anything, it should be this:

``` perl
    "module"                => [ "fips" ],
```

mattcaswell replied:
As an aside the "no-modules" option appears to be undocumented.

levitte replied:
Very good point.  I'll get on it.

levitte commented on INSTALL:
```diff
@@ -394,6 +394,9 @@
 
+  no-fips
+                   Don't compile the FIPS module
+
```
I'm not sure it should generate the `OPENSSL_NO_FIPS` macro, though.  That will just add confusion, 'cause that's not a library option.

There's this section in `Configure` that has a selection of config options that shouldn't end up as C macros of that form, `'fips'` should be added there:

https://github.com/openssl/openssl/blob/e321ba28e89d40ab750e353d5bd7fc405e9c2761/Configure#L1224-L1226

mattcaswell replied:
> I'm not sure it should generate the OPENSSL_NO_FIPS macro, though. That will just add confusion, 'cause that's not a library option.

I guess I don't understand that argument. It's something that we want to be able to write conditional code on (as evidenced by the fact I've used it already in the tests) - so why would we suppress the macro generation?

mattcaswell replied:
Hmmm. Thinking about this more - I guess this is about whether it is appropriate to have the macro in opensslconf.h (a public header file). Is that it?

levitte replied:
Yes, exactly

levitte commented on test/evp_extra_test.c:
```diff
@@ -1207,6 +1250,10 @@ int setup_tests(void)
     ADD_ALL_TESTS(test_invalide_ec_char2_pub_range_decode,
                   OSSL_NELEM(ec_der_pub_keys));
 #endif
+#ifdef OPENSSL_NO_FIPS
```
However, you can always do the same I did for `provider_internal_test` here, in `build.info`:

```
    IF[{- $disabled{fips} -}]
      DEFINE[evp_extra_test]=NO_FIPS_MODULE
    ENDIF
```

and use that as a guard...

mattcaswell replied:
It seems strange to me to deliberately suppress the OPENSSL_NO_FIPS macro - only to then create a new one instead.

levitte replied:
The difference is that this is telling this specific test app if the fips module is available or not, i.e. entirely based on what's being built here and now, rather than on something that looks like a library feature.

levitte replied:
BTW, the condition should be this:

```
    IF[{- $disabled{fips} || !$target{dso_scheme} -}]
```

The lack of a DSO scheme doesn't necessarily mean that the module isn't built, it only means that we can't load it.

levitte replied:
Speaking of macros, this had me rethink that macro name I used for the provider tests.  See #8664

paulidale replied:
I thought we'd decided on `FIPS_MODE` as the macro to be defined.

levitte replied:
You may want to read https://www.openssl.org/docs/OpenSSL300Design.html#code-maintenance again.  `FIPS_MODE` isn't meant to be used to say if the fips provider is available or not.

slontis commented:
So this may just be a transitional problem, but I noticed something strange today (its not currently an issue I need adressing - but it might be worth noting).
I loaded a test provider and then unloaded it. After this I tried using HMAC SHA256 and it fell over when trying to fetch the digest. Changing it to use SHA224 worked (since that uses the deprecated code path).
Loading and unloading the provider after doing a HMAC-SHA256 worked properly.

levitte commented:
My hunch is that's because as soon as any provider has been loaded, fallbacks aren't available as such any more in the lifetime of that process. One might argue that it should still fall back to legacy code paths, but considering that's only an interim thing, I don't think we should waste any effort there.

Bottom line, this behavior is along the lines of what has been discussed earlier, that as soon as an application starts specifying providers explicitly, be it by loading directly or via configuration, then it must also specify the default provider among them, otherwise it's not available.

mattcaswell commented:
Fix-up commits pushed addressing @levitte's comments.

mattcaswell commented:
Grrr...travis failure still. Pushed an update that hopefully fixes it. Also, offline, @levitte said to me that he had changed his mind about where the sources for the module should be specified. I've changed it back to the way I used to have it - which is what I think you intended.

levitte commented:
You understood me perfectly, Matt.  The reasoning is that for large number of source files, it's better to keep the specs together, and close to the source, the same way we currently specify what goes into libcrypto.  It also permits us to use common lists instead of copying so much, i.e. something like this:

```
{- our $common_source = join(' ', qw(foo1.c foo2.c)); -}
SOURCE[../.../default]=whatever_def.c {- $common_source -}
SOURCE[../.../fips]={- $common_source -} whatever_fips.c
```

Something like that...

mattcaswell commented:
Including mem_dbg.c turns out to be not so easy...

levitte commented:
Yeah, I've noticed...  I was surprised how little you were sucking into the FIPS module!  The whole thing is quite entangled.

A solution might be to turn off mdebug when building the FIPS module.  It's not like the data collected will do anyone anything good anyway, unless you plan on calling its leak checking function at teardown...

mattcaswell commented:
> I was surprised how little you were sucking into the FIPS module! The whole thing is quite entangled.

Much more will come in later (I've already have some work to do this) - but I wanted to keep it minimal for this PR

mattcaswell commented:
Another attempt at getting travis to go through successfully. I've avoided pulling in mem_dbg.c because that wants to suck in a whole load of dependencies that I want to avoid at this stage. Instead I've just disabled crypto-mdebug within the context of the FIPS module (but not elsewhere) for now. I also found and fixed a mem-leak in the test while doing this!

mattcaswell commented:
Woop! All the CIs passed!

mattcaswell commented:
Pushed. Thanks!!!
